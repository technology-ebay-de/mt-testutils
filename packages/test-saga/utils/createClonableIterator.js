/**
 * createClonableIterator
 *
 * Allows to create a generator function that has an added `clone` property, that will return a
 * cloned iterator with the same state as the original iterator.
 *
 * Changing the state of the cloned iterator will not affect the original iterator's state.
 *
 * Example:
 *
 *   function* myGen(someArg) {
 *       yield 1;
 *       yield 2;
 *       yield 3;
 *   }
 *
 *   // cloning a generator
 *   const iterator = createClonableIterator(myGen, 'someArgValue');
 *   const clonedIterator = iterator.clone();
 */

const createClonableIterator = (getRootIterator, history = []) => {
    const iterator = getRootIterator();
    const next = iterator.next.bind(iterator);

    history.forEach((key, value) => {
        next(value);
    });

    iterator.next = input => {
        history.push(input);
        return next(input);
    };

    iterator.clone = () => createClonableIterator(getRootIterator, history.slice(0));

    return iterator;
};

export default (generator, generatorArgsArray = []) => {
    const getRootIterator = () => generator.apply(null, generatorArgsArray);
    return createClonableIterator(getRootIterator);
};
