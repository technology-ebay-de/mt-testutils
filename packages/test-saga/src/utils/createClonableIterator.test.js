import createClonableIterator from './createClonableIterator';

function* myGenerator(one) {
    yield one;
    yield 2;
    yield 3;
}

describe('When creating a clonable iterator', () => {
    let iterator;
    beforeEach(() => {
        iterator = createClonableIterator(myGenerator, [1]);
    });
    describe('the returned iterator', () => {
        it('has a `next` method', () => {
            iterator.next.should.be.a('function');
        });
        it('has a `clone` method', () => {
            iterator.clone.should.be.a('function');
        });
        it('can be iterated on', () => {
            const nextSteps = [iterator.next(), iterator.next(), iterator.next(), iterator.next()];
            expect(nextSteps).toMatchSnapshot();
        });
    });
    describe('the clone of the resulting iterator', () => {
        let clonedIterator;
        beforeEach(() => {
            clonedIterator = iterator.clone();
        });
        it('has a `next` method', () => {
            clonedIterator.next.should.be.a('function');
        });
        it('has a `clone` method', () => {
            clonedIterator.clone.should.be.a('function');
        });
        it('can be iterated on', () => {
            const nextSteps = [
                clonedIterator.next(),
                clonedIterator.next(),
                clonedIterator.next(),
                clonedIterator.next()
            ];
            expect(nextSteps).toMatchSnapshot();
        });
        it('should behave independently from original iterator', () => {
            const nextSteps = [iterator.next(), iterator.next(), iterator.next(), iterator.next()];
            expect(nextSteps).toMatchSnapshot();
        });
    });
    describe('and cloning an already started iterator', () => {
        let clonedIterator;
        beforeEach(() => {
            iterator.next();
            clonedIterator = iterator.clone();
        });
        describe('the two iterators', () => {
            it('iterate independently from each other', () => {
                iterator.next().value.should.equal(2);
                clonedIterator.next().value.should.equal(2);
            });
        });
    });
});
