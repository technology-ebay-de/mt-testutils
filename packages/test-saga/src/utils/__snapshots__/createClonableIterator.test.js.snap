// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`When creating a clonable iterator the clone of the resulting iterator can be iterated on 1`] = `
Array [
  Object {
    "done": false,
    "value": 1,
  },
  Object {
    "done": false,
    "value": 2,
  },
  Object {
    "done": false,
    "value": 3,
  },
  Object {
    "done": true,
    "value": undefined,
  },
]
`;

exports[`When creating a clonable iterator the clone of the resulting iterator should behave independently from original iterator 1`] = `
Array [
  Object {
    "done": false,
    "value": 1,
  },
  Object {
    "done": false,
    "value": 2,
  },
  Object {
    "done": false,
    "value": 3,
  },
  Object {
    "done": true,
    "value": undefined,
  },
]
`;

exports[`When creating a clonable iterator the returned iterator can be iterated on 1`] = `
Array [
  Object {
    "done": false,
    "value": 1,
  },
  Object {
    "done": false,
    "value": 2,
  },
  Object {
    "done": false,
    "value": 3,
  },
  Object {
    "done": true,
    "value": undefined,
  },
]
`;
